XBeeModuleSim.py

```diff
@@ -0,0 +1,197 @@
+
+    def recieved_from_rocket(self, data):
+        """
+        :brief: All data incoming from the rocket should be passed into this method for processing.
+        :param data: bytearray of data recieved.
+        """
+        self._rocket_rx_queue_packed.put(data)
+
+    def __init__(self):
+        """
+        Constructor.
+        In addition to constructing this, if any useful work is to be done then the rocket_callback and ground_callback attributes should be set - by default they are simply no-op functions.
+        """
+        self._rocket_rx_queue = self._unpack(self._rocket_rx_queue_packed)
+
+        # Callbacks should be of form callback(data), where data is a bytearray.
+        # Making a no-op callback to prevent need for checking if callback exists.
+        def nop_callback(data):
+            return
+
+        """Is called whenever data needs to be sent to the rocket through SIM - this should be the SIM send() function. The callback should be thread safe."""
+        self.rocket_callback = nop_callback
+
+        """Is called whenever data recieved from the rocket needs to be sent to the rest of the ground station code. The callback should be thread safe."""
+        self.ground_callback = nop_callback
+
+        self._running_lock = Lock()
+        self._running = True
+
+        # Queues are IO bound.
+        self._rocket_rx_thread = Thread(
+            target=self._run_rocket_rx, name="xbee_sim_rocket_rx"
+        )
+        self._ground_rx_thread = Thread(
+            target=self._run_ground_rx, name="xbee_sim_ground_rx"
+        )
+        self._rocket_rx_thread.start()
+        self._ground_rx_thread.start()
+
+    def shutdown(self):
+        """ 
+        Shuts down everything and kills the internal threads.
+        Callbacks will not be interrupted/passed corrupt data.
+        """
+        with self._running_lock:
+            self._running = False
+
+        # Fill up the queues with garbage to unblock the threads on input
+        self._ground_rx_queue_packed.put(b"\0")
+        self._rocket_rx_queue_packed.put(b"\0")
+
+        self._rocket_rx_thread.join()
+        self._ground_rx_thread.join()
+
+    @property
+    def running(self):
+        with self._running_lock:
+            return self._running
+
+    def _unpack(self, q):
+        """
+        Helper generator that unpacks the iterables in a given queue.
+        """
+        while self.running:
+            arr = q.get()
+            for i in arr:
+                yield i
+        sys.exit()
+
+    def _run_rocket_rx(self):
+        """Process the incoming rocket data queue."""
+        while True:
+            self._parse_API_frame()
+
+    def _run_ground_rx(self):
+        """Process the ground station data queue."""
+        while self.running:
+            reserved = b"\xff\xfe"
+            rx_options = b"\x02"
+            data = self._ground_rx_queue_packed.get()
+            self.rocket_callback(
```

Comment:
Was not aware of `SendThread`; was trying to keep it as drop-in as possible.
I mean strictly speaking there's a few minor changes that's done by SendThread - it converts keywords into their equivalent commands and checks for an XBee timeout exception - but nothing that can't be overcome. 
I'm not familiar with how objects get passed around so I'm not sure whether this would work, but the solution that comes to mind is convert the SendThread queue reader into a (series of?) generator(s), which somehow gets passed to the XBeeSimulator. Something that throws a wrench in is that this needs to be conditional on the connection type (SIM, Debug, Serial).

Reply:
I think its as simple as putting the contents of _run_ground_rx in send_to_rocket. No thread, no queue, send immediately. I don't see why  that wouldn't work?

Reply:
I remember that doing it seemed like a good idea at the time but I'm not sure if that was due to an actual reason or just being sleep deprived
I'll have to wait until I can figure out what I'm doing wrong with PyTest so I can test it, but once that's done I'll fix it

Reply:
Yep fixed - thanks for noticing that